File: ohara-kafka/src/main/scala/com/island/ohara/kafka/TopicCreator.scala
package com.island.ohara.kafka
import org.apache.kafka.common.config.TopicConfig

import scala.concurrent.duration._

/**
  * a helper class used to create the kafka topic.
  * all member are protected since we have to implement a do-nothing TopicCreator in testing.
  */
abstract class TopicCreator {
  private[this] var numberOfPartitions: Int = 1
  private[this] var numberOfReplications: Short = 1
  private[this] var options: Map[String, String] = Map.empty
  private[this] var timeout: Duration = 10 seconds

  def numberOfPartitions(numberOfPartitions: Int): TopicCreator = {
    this.numberOfPartitions = numberOfPartitions
    this
  }

  def numberOfReplications(numberOfReplications: Short): TopicCreator = {
    this.numberOfReplications = numberOfReplications
    this
  }

  def options(options: Map[String, String]): TopicCreator = doOptions(options, true)

  /**
    * Specify that the topic's data should be compacted. It means the topic will keep the latest value for each key.
    * @return this builder
    */
  def compacted(): TopicCreator = {
    doOptions(Map(TopicConfig.CLEANUP_POLICY_CONFIG -> TopicConfig.CLEANUP_POLICY_COMPACT), false)
    this
  }

  /**
    * Specify that the topic's data should be deleted. It means the topic won't keep any data when cleanup
    * @return this builder
    */
  def deleted(): TopicCreator = {
    doOptions(Map(TopicConfig.CLEANUP_POLICY_CONFIG -> TopicConfig.CLEANUP_POLICY_DELETE), false)
    this
  }

  private[this] def doOptions(options: Map[String, String], overwrite: Boolean): TopicCreator = {
    if (this.options == null || overwrite) this.options = options
    else {
      this.options
        .filter {
          case (k, _) => options.contains(k)
        }
        .foreach {
          case (k, v) =>
            if (options(k) != v) throw new IllegalArgumentException(s"conflict options! previous:$v new:${options(k)}")
        }
      this.options ++= options
    }
    this
  }

  def timeout(timeout: Duration): TopicCreator = {
    this.timeout = timeout
    this
  }
  import TopicCreator._
  def create(name: String): Unit = doCreate(Request(name, numberOfPartitions, numberOfReplications, options, timeout))

  protected def doCreate(request: Request): Unit
}

object TopicCreator {
  case class Request(name: String,
                     numberOfPartitions: Int,
                     numberOfReplications: Short,
                     options: Map[String, String],
                     timeout: Duration)
}

File: ohara-kafka/src/main/java/com/island/ohara/kafka/TopicCreator.java
/*
 * Copyright 2019 is-land
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.island.ohara.kafka;

import java.time.Duration;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import org.apache.kafka.common.config.TopicConfig;

/**
 * a helper class used to create the kafka topic. all member are protected since we have to
 * implement a do-nothing TopicCreator in testing.
 */
public abstract class TopicCreator {
  protected int numberOfPartitions = 1;
  protected short numberOfReplications = 1;
  protected Map<String, String> options = new HashMap<>();
  protected Duration timeout = Duration.ofSeconds(10);

  public TopicCreator numberOfPartitions(int numberOfPartitions) {
    this.numberOfPartitions = numberOfPartitions;
    return this;
  }

  public TopicCreator numberOfReplications(short numberOfReplications) {
    this.numberOfReplications = numberOfReplications;
    return this;
  }

  public TopicCreator options(Map<String, String> options) {
    doOptions(options, true);
    return this;
  }

  /**
   * Specify that the topic's data should be compacted. It means the topic will keep the latest
   * value for each key.
   *
   * @return this builder
   */
  public TopicCreator compacted() {
    doOptions(
        Collections.singletonMap(
            TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_COMPACT),
        false);
    return this;
  }
  /**
   * Specify that the topic's data should be deleted. It means the topic won't keep any data when
   * cleanup
   *
   * @return this builder
   */
  public TopicCreator deleted() {
    doOptions(
        Collections.singletonMap(
            TopicConfig.CLEANUP_POLICY_CONFIG, TopicConfig.CLEANUP_POLICY_DELETE),
        false);
    return this;
  }

  private TopicCreator doOptions(Map<String, String> options, boolean overwrite) {
    if (this.options == null || overwrite) {
      this.options = new HashMap<>(options);
    } else {
      this.options
          .entrySet()
          .stream()
          .filter(x -> options.containsKey(x.getKey()))
          .forEach(
              x -> {
                if (!options.get(x.getKey()).equals(x.getValue()))
                  throw new IllegalArgumentException(
                      String.format(
                          "conflict options! previous:%s new:%s",
                          x.getValue(), options.get(x.getKey())));
              });

      this.options.putAll(options);
    }
    return this;
  }

  public TopicCreator timeout(Duration timeout) {
    this.timeout = timeout;
    return this;
  }

  public abstract void create(String name);
}
